### DISCUSSION TITLE
<Dialog /> / <Portal />  custom root element

### DISCUSSION THREAD

**Discussion by michael-land** [tailwindlabs/headlessui#666] [Ideas]
### What package within Headless UI are you using?

@headlessui/react

### What version of that package are you using?

1.0.0

### What browser are you using?

chrome

### Reproduction repository

I'm rendering my react app under shadow root and all my css are scoped inside the shadow root.

I want the `<Dialog />` or the `<Portal />` insert inside the shadow root instead of the document.body. It would be nice we can pass custom root element for  `<Dialog />` / `<Portal />`.

![image](https://user-images.githubusercontent.com/33362998/114949370-80a99a80-9e16-11eb-82f2-2b48505e9121.png)

![image](https://user-images.githubusercontent.com/33362998/114949731-370d7f80-9e17-11eb-8f73-e12fdadcbe69.png)

**Comment by SigurdMW** [comment]
Any update on this? This is a feature we would like as well 😄

**Comment by zawilliams** [comment]
Running into this same issue. Custom root element would be great.

**Comment by rikkipitt** [comment]
Same here. I think I'm running into a similar thing...

**Comment by MarcelMalik** [comment]
Same issue here. Workaround: must use a custom modal implementation :/

**Comment by xylophonehero** [comment]
Would be great to have this feature. My custom Modal implementation isn't as good as this one 😢. Using styled components and my CSS is scoped to the top of my shadow too. A custom element for where the portal should be rendered would be a great implementation!

**Comment by dcastil** [comment]
I also have the use case where I don't want to use a portal because the dialog needs to be positioned based on the element close to which the React component is rendered. The dialog takes in most of the screen space, the background gets blurred while the dialog is open and I want to trap the focus in the dialog, so using a Popover wouldn't be the right choice here.

I need to use a [different library](https://www.npmjs.com/package/react-focus-on) for this although the functionality is basically the same as the Headless UI dialog.

Would be awesome if Headless UI could support not using a portal at all! ❤️

**Comment by yaakov123** [comment]
I have an unbelievably hacky solution, but it works (sort of).

Paste the following code as early as possible (index.ts or something)

```javascript
const elementById = Document.prototype.getElementById;
Document.prototype.getElementById = function (elementId: string) {
  if (elementId === "headlessui-portal-root") {
    return root.element; // This is the root element which exists inside the Shadow DOM
  }
  return elementById.call(this, elementId);
};
```

The issue now is that clicking anywhere closes the Modal

**Reply by yaakov123** [reply]
This issue can be solved by removing the @close listener from the Dialog component and manually placing it on the overlay

**Comment by yaakov123** [comment]
I've written a more complete solution here -  https://github.com/tailwindlabs/headlessui/discussions/874#discussioncomment-1894010

**Comment by jtomaszewski** [comment]
Dialog should allow for customizing what's/where's the portal root node.

This will solve problems like:

- this one
- https://github.com/tailwindlabs/headlessui/issues/998

Until then, devs have to choose a different Dialog implementation, or go around with hacks (like replacing the `Document.prototype.getElementById` or creating `<div id="headlessui-portal-root" />` by themselves and making sure it's recreated between dialog mounts/dismounts).

@RobinMalfait tell me that you'd be open to this, and I can raise a PR.

**Reply by jtomaszewski** [reply]
IMO the best API would be to store the reference to the portal root node in a context, so that one (optionally!) can replace it, sth like this:

```tsx
<React.Fragment>
  <div id="my-custom-portal-root" ref={portalRootRef} />
  <PortalRootContext.Provider value={portalRootRef}>
    <Dialog ... />
  </PortalRootContext.Provider>
</React.Fragment>
```

where `const PortalRootContext = React.createContext<{ current: HTMLElement }>();` .

**Reply by jtomaszewski** [reply]
P.S. For now, as a simplier workaround, you can just do

```tsx
<React.Fragment>
  <YourAppThatUsesDialogs ... />
  
  {/* Workaround for https://github.com/tailwindlabs/headlessui/discussions/666 */}
  <div id="headlessui-portal-root">
    {/* It needs at least one child, so that HeadlessUI doesn't remove this portal root workaround
        ( https://github.com/tailwindlabs/headlessui/blob/main/packages/@headlessui-react/src/components/portal/portal.tsx#L84 ) */}
    <div />
  </div>
</React.Fragment>
```

**Reply by lushkant** [reply]
Thanks a ton @jtomaszewski that does it 🚀

**Reply by yassilah** [reply]
thanks a lot, works for me too 🙂

**Reply by okpedro** [reply]
> P.S. For now, as a simplier workaround, you can just do
> 
> ```tsx
> <React.Fragment>
>   <YourAppThatUsesDialogs ... />
>   
>   {/* Workaround for https://github.com/tailwindlabs/headlessui/discussions/666 */}
>   <div id="headlessui-portal-root">
>     {/* It needs at least one child, so that HeadlessUI doesn't remove this portal root workaround
>         ( https://github.com/tailwindlabs/headlessui/blob/main/packages/@headlessui-react/src/components/portal/portal.tsx#L84 ) */}
>     <div />
>   </div>
> </React.Fragment>
> ```

I found that if I set it to `<div/>` or an empty `<div></div>`, it doesn't work. Even if I put whitespace in there or fill it with content and hide it, it won't work. But if I put a single visible character, like `<div>i</div>`, it works. Otherwise, the modal opens and closes in a flash. I'm scratching my head trying to figure out why that matters.

EDIT: Figured out `<div/>` only doesn't work if the `<Dialog>` component is set to fixed instead of relative, in which case the `<div>` needs some visible content.

**Reply by steph-crown** [reply]
When I did this, clicking on the dialog itself closes the dialog. @jtomaszewski please, do you have any fix for that.

**Reply by paweltatarczuk** [reply]
@Steph-crown I encountered a similar issue and had to use a different workaround in my case. The dialog was not interactive because of an `inert` tag that was added to an ancestor of the `<div id="headlessui-portal-root">`. Since I was using this workaround only for scoped styles, I decided to solve it with the `postcss-prefix-selector` plugin instead.

**Reply by sturmenta** [reply]
@Steph-crown my fix was (suggested by @thevisioner [here](https://github.com/tailwindlabs/headlessui/discussions/666#discussioncomment-6183443))

remove the `inert` tag from the root component that I was using:

```ts
import { useEffect } from "react";

export const useRemoveInert = ({ isModalOpen }: { isModalOpen: boolean }) => {
  useEffect(() => {
    if (isModalOpen) {
      // timeout needed to make sure inert is removed after dialog is opened
      setTimeout(() => {
        // disable inert to make dialog interactive -> https://github.com/tailwindlabs/headlessui/discussions/666#discussioncomment-6183443
        const rootNode = document.getElementById("your-root-id");
        if (rootNode) rootNode.inert = false;
      }, 100);
    }
  }, [isModalOpen]);
};

```

**Comment by sbarbat** [comment]
For some reason non of the solutions here helped me, so listening to changes in the DOM and then moving the portal root to wherever I wanted worked for me:

```typescript
// Just create the container and shadow root
let container = document.getElementById('some-container');
if (!container) {
  container = document.createElement('div');
  container.id = 'some-container';
  document.body.appendChild(container);
}
const shadow = container.shadowRoot ?? container.attachShadow({ mode: 'open' });

/**
 * Every time a new element `headlessui-portal-root` is added to the DOM,
 * it will place it into the shadow body inside `some-container`
 */
new MutationObserver((records: MutationRecord[]) => {
  const record: MutationRecord | undefined = records
    .filter((record) => (record.target as HTMLElement).id === 'headlessui-portal-root')
    .pop();

  if (!record) return;
  shadow.append(record.target);
}).observe(document, { childList: true, subtree: true });
```

**Reply by yaakov123** [reply]
The issue over here is that the click to hide functionality will no longer work with this solution. For example, with the Dialog element, with this solution when you click on the Dialog itself, the Dialog closes.

**Reply by sbarbat** [reply]
Yes, that something I'm facing, do you know why this might be happening?

**Reply by yaakov123** [reply]
You can take a look at my solution here for a working fix: https://github.com/tailwindlabs/headlessui/discussions/874#discussioncomment-1894010

**Reply by animber0** [reply]
@yaakov123  This one works for me, but can you help me understand how to convert this solution to yours, or what the fundamental differences are?

**Reply by RomainMazB** [reply]
Because the `inert` attribute is added to the HTML element hosting the Vue app (probably `#app`)
A simpler solution is to move the `div#headlessui-portal-root` outside of your Vue app.

[As mentioned in the vue doc](https://vuejs.org/guide/built-ins/teleport#basic-usage) it won't affect the Portal behavior:
> Sometimes a part of a component's template belongs to it logically, but from a visual standpoint, it should be displayed somewhere else in the DOM, **_perhaps even outside of the Vue application._**
```html
<body>
    <div id="app"></div> <!-- Will receive the inert attribute -->

    <div id="headlessui-portal-root">
        <div></div>
    </div>
</body>
```

**Reply by leizzo** [reply]
@yaakov123 if the dialogue is set to open by default, your solution does not work correctly. but as for my problem, if the opened dialogue is closed and you want to open it again, it does not work. I could not solution for this :/

**Comment by pfcodes** [comment]
Forcing a single portal that is unmodifiable is a pretty big architectural flaw imo. We should be able to use our own like Radix UI does https://www.radix-ui.com/docs/primitives/components/dialog#custom-portal-container

**Comment by curtgrimes** [comment]
I really need this for my use case.

**Comment by loicmagne** [comment]
🙏

**Reply by alifatic** [reply]
i still got this issue in   "@headlessui/react": "^1.6.6", please any fix, other solution does not fix this

**Comment by lukeclifton** [comment]
I have the same issue here. As a work around i am doing the following:

``` 
<div id="headlessui-portal-root" class="tailwind-important-class-or-id">
    <div id="react-app-root-id"></div>
</div>
```

**Comment by web-programmer-here** [comment]
any updates on this?

**Comment by franco14lorenzo** [comment]
For those who happen to them in Next.js and the new Font component , this was my solution:

- In the _app.jsx file, add a div to wrap the page component, it must be a child element of the element that is applying the font variable

`<div className={`${montserrat.variable} font-sans`}>
            <div id="headlessui-portal-root">
              <Component {...pageProps} err={err} />
            </div>
          </div>`

**Reply by Maisaan** [reply]
This works, to answer OP's question, adding the id of `headlessui-portal-root` to your main or root div results in the portal being rendered inside of that element.

**Reply by Cretezy** [reply]
To add to this, `Component` should not be rendered inside. You want the `headlessui-portal-root` to be the last item ideally, to minic how it behaves if it's not manually set.

Full example (in your `pages/_app`):
```tsx
const font = Noto_Sans({
  variable: "--next-font",
  subsets: ["latin"],
  weight: ["400"],
});

export default function App({
  Component,
  pageProps,
}: AppProps) {
  return (
    <div
      className={clsx(
        // The important line:
        font.variable,
        // className is optional. It will set the `font-family`
        // Alternatively, add the snippet below this one inside your tailwind.config.json 
        font.className,
      )}
    >
      <Component {...pageProps} />

      {/* Workaround for https://github.com/tailwindlabs/headlessui/discussions/666, to allow font CSS variable */}
      <div id="headlessui-portal-root">
        {/* Important that a div is inside the portal root */}
        <div />
      </div>
    </div>
  );
}
```

For `tailwind.config.js`:

```js
module.exports = {
  // ...,
  theme: {
    fontFamily: {
      sans: [
        // Important line, uses the font family from Next Font
        "var(--next-font)",
        // Your fallback. This was my default from Tailwind
        "ui-sans-serif",
        "system-ui",
        "-apple-system",
        "BlinkMacSystemFont",
        '"Segoe UI"',
        "Roboto",
        '"Helvetica Neue"',
        "Arial",
        '"Noto Sans"',
        "sans-serif",
        '"Apple Color Emoji"',
        '"Segoe UI Emoji"',
        '"Segoe UI Symbol"',
        '"Noto Color Emoji"',
      ],
    },
  },
};
```

**Reply by AdrianMachado** [reply]
For anyone using the new App Directory in Next 13, this approach will work for fixing issues with tailwind CSS custom variables not being available at runtime in your headless components. For example, I use `Dialog` and `react-modal-hook` to render a modal, but the mounting point was too high for tailwind to apply properly. I moved the mounting point within a lower hierarchy `layout.tsx` file and it works again!

**Reply by peyronoscar** [reply]
I get the following error when trying this workaround: `There are no focusable elements inside the <FocusTrap />`. Any ideas why?

**Reply by franco14lorenzo** [reply]
> I get the following error when trying this workaround: `There are no focusable elements inside the <FocusTrap />`. Any ideas why?

I recently had the same problem and what I did was remove the div and change the way   load the font:

In _app.js:
```
<style jsx global>{`:root { --font-montserrat: ${montserrat.style.fontFamily};}`}</style>
<Component {...pageProps} err={err} />

```

**Reply by bhatvikrant** [reply]
This works! Thanks @franco14lorenzo

**Reply by sipec** [reply]
Workaround no longer works as of headless ui 2.1

**Comment by bpolaszek** [comment]
Just got trapped into that same issue, by trying to introduce dynamic CSS variables. In Vue, they can't be bound to `<body>` (AFAIK) so classes that depend on them must be used within the root component and not inside one of its siblings.

```vue
<script setup>
// Root.vue
const primaryColor = computed(() => ({
  50: '#eff9ff',
  100: '#def2ff',
  200: '#b6e7ff',
  300: '#75d7ff',
  400: '#2cc3ff',
  500: '#00A0E6',
  600: '#0089d4',
  700: '#006dab',
  800: '#005c8d',
  900: '#064d74',
  950: '#04304d',
})); // Computed, because values can change at runtime (changing app theme)
</script>

<style lang="scss">
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer base {
  * {
    --color-primary-50: v-bind(primaryColor[50]);
    --color-primary-100: v-bind(primaryColor[100]);
    --color-primary-200: v-bind(primaryColor[200]);
    --color-primary-300: v-bind(primaryColor[300]);
    --color-primary-400: v-bind(primaryColor[400]);
    --color-primary-500: v-bind(primaryColor[500]);
    --color-primary-600: v-bind(primaryColor[600]);
    --color-primary-700: v-bind(primaryColor[700]);
    --color-primary-800: v-bind(primaryColor[800]);
    --color-primary-900: v-bind(primaryColor[900]);
    --color-primary-950: v-bind(primaryColor[950]);
  }
}
</style>

<template>
  <RouterView />
</template>
```

```js
// tailwind.config.js
module.exports = {
  theme: {
    extend: {
      colors: {
        primary: {
          50: 'var(--color-primary-50)',
          100: 'var(--color-primary-100)',
          200: 'var(--color-primary-200)',
          300: 'var(--color-primary-300)',
          400: 'var(--color-primary-400)',
          500: 'var(--color-primary-500)',
          600: 'var(--color-primary-600)',
          700: 'var(--color-primary-700)',
          800: 'var(--color-primary-800)',
          900: 'var(--color-primary-900)',
          950: 'var(--color-primary-950)',
        }
      },
    },
  },
};
```

```vue
<script setup>
// PrimaryButton.vue
</script>
<template>
  <button class="bg-primary-700 text-white">
    <slot />
  </button>
</template>
```

Rendering `PrimaryButton.vue`: 
- Under `Root.vue`: 👌 (white text on `#006dab` background)
- Within a `Dialog` component: 💥  (white text on transparent background)

Thanks a lot @okpedro for [saving my day](https://github.com/tailwindlabs/headlessui/discussions/666#discussioncomment-3449763)!

**Comment by thevisioner** [comment]
Just wanted to mention that `<div id="headlessui-portal-root">`  helps with Next.js font scoping. Also added a little helper just before the closing `</Dialog>` to remove the 'inert' attribute and keep the dialog interactive:

```tsx
function DisableInert() {
  useEffect(() => {
    const rootNode = document.getElementById("__next");
    if (rootNode) {
      rootNode.inert = false;
    }
  }, []);
  return null;
}
```

**Reply by xXDeonoXx** [reply]
omg, thank you, that was just what i needed but i didn't though how simple it could have been

**Comment by ScripterSugar** [comment]
Why is this issue not fixed after 3 years of initial raise? any particular reason?

This flaw significantly affects the main concept of headless UI - fully customizable, the modal component is completely useless for my usecase because of this ;*(

**Reply by axe312ger** [reply]
Have to agree :(

Edit:

Well, there is id="headlessui-portal-root" which tells headless ui where to put the portal. Works fine for me! :)

**Comment by Welfordian** [comment]
Was there any solution found for this problem? :-(

**Comment by TrustyTechSG** [comment]
I facing the Dialog display issue when rendering Dialog in iframe (using react-frame-component), the Dialog will always attach to the outer window instead of the iframe window.

**Comment by heronoa** [comment]
The way I solve it was to set a Modal on the my template root (the first element that will render and use context to open and close it. TL;DR I think I create my own modal but still use Transitions from headlessUI to animate it

**Comment by wade-wojcak-vib** [comment]
Wanted to chime in with my use case where I stumbled on this issue: 

I am using Storybook + React and was testing out a Dialog. I have a wrapper around my Stories at the preview level which applies my custom font. This works fine for the baseline rendering of a Story, but when a Dialog Portal is created it is created outside of that wrapper and thus does not retain the font. 

I can fix this for an initial Dialog opening by defining a div with id = 'headlessui-portal-root' within my Stories rendering function, but this gets wiped out after you close and reopen the Dialog a second time. Given this issue is just in a testing environment I will probably just overlook it for now, but it would be nice to have a way to define a permanent portal root if necessary.

**Comment by SampsonCrowley** [comment]
this makes it impossible to have more than one root (i.e. separate apps on the same page)

It also makes it soe that you can't guarantee certain dialogs are at a higher stacking context than others

**Comment by ChaseIngebritson** [comment]
Chiming in with another workaround for usage in Storybook using the [removeInert hook](https://github.com/tailwindlabs/headlessui/discussions/666#discussioncomment-6898923) from @sturmenta and the [storybook tips](https://github.com/tailwindlabs/headlessui/discussions/666#discussioncomment-7721293) from @wade-wojcak-vib:

```jsx
export const Default: Story = {
  args: {
    open: false
  },
  render: function Render(args) {
    const [open, setOpen] = useState(args.open)

    // This can be extracted into a hook, but for the sake of example it's not
    useEffect(() => {
      if (open) {
        setTimeout(() => {
          const rootNode = document.getElementById('storybook-root')
          if (rootNode) rootNode.inert = false
        }, 100)
      }
    }, [open])

    return (
      <>
        <div id="headlessui-portal-root" />
        <MeshProviderMock mocks={globalQueryMocks}>
          <SaturateCacheMock>
            <ButtonPrimary label="Toggle" onClick={() => setOpen(!open)} />
            {/* This EditorDialog component is a thin wrapper for the headlessui dialog component and is not rendered until the button is clicked */}
            {open && <EditorDialog open={open} />}
          </SaturateCacheMock>
        </MeshProviderMock>
      </>
    )
  }
}

```

Note that the headlessui portal root component needs to exist in the dom before the headlessui dialog component has been rendered, otherwise it won't detect that a root already exists in the ownerDocument

**Reply by johnnyshankman** [reply]
-1 this only works for me on the first opening. if i close the dialog, it removes my predefined headless portal root from the dom and then proceeds to do default behavior upon opening the dialog a second time.

**Comment by AshleySchaeffer** [comment]
Not ideal, but in React I was trying to have notifications appear over all other elements. This issue caused me problems. I just did this:

https://react.dev/reference/react-dom/createPortal#rendering-a-modal-dialog-with-a-portal

It creates another portal which renders as expected.

**Reply by KrishEnacton** [reply]
Can you provide me a full example of it?

**Reply by AshleySchaeffer** [reply]
https://codesandbox.io/p/sandbox/6c6clm

If you click the button in the top right of the code snippets on the docs, you'll see a full example hosted on codesandbox. 

I don't fully understand it, but it looks like portals dynamically create elements at the document root. I just defined a portal in the component and it worked fine, even rendering after a modal component rendered within another portal. 

As the notification component renders after the modal, it's portal element renders above the modal.

**Comment by johnnyshankman** [comment]
This is so silly that we can't predefine where the portal goes to using a prop.
```ts
<div id="headlessui-portal-root" />
```

You can try putting this in your DOM but it will only hook into it one time. Once you close the Dialog, it will remove this element and then reinject it in the default spot. So silly.

Took me twenty minutes reading through this to find my solve:

```ts
  {/* It needs at least one child, so that HeadlessUI doesn't remove this portal root workaround
        ( https://github.com/tailwindlabs/headlessui/blob/main/packages/@headlessui-react/src/components/portal/portal.tsx#L84 ) */}
 ```

**Reply by johnnyshankman** [reply]
Oh great but now the z-indexing is completely screwed so I cannot click on anything inside my dialogs. I give up what an awful library. Seems like an obvious thing you'd want to control. I don't just want portals opening up at the root of the document.

**Reply by AshleySchaeffer** [reply]
> Oh great but now the z-indexing is completely screwed so I cannot click on anything inside my dialogs. I give up what an awful library. Seems like an obvious thing you'd want to control. I don't just want portals opening up at the root of the document.

See my solution. It worked for me. Might work for you.

**Comment by GianlucaGuarini** [comment]
I have just figured out that with `<div id="headlessui-poral-root">` workaround the dialog is broken. The `inert` attribute is not properly rendered through the DOM hierarchy (see also https://github.com/tailwindlabs/headlessui/issues/3632).

Please make sure to upvote this discussion in order to let the maintainers accept PRs to fix this annoying issue https://github.com/tailwindlabs/headlessui/discussions/3563

**Comment by edda** [comment]
I have the same problem that I need my portal root to be rendered inside the shadow root. Not being able to set a custom portal root element is causing me so much extra work because I now have to manually take care of positioning everywhere, it's very frustrating!

Alternatively I would also take not having the headlessui portal forced on me if using anchor positioning. If I could use `anchor` and set `portal={false}` at the same time and then create my own portal around the component that would also work for me.

**Comment by ercgrat** [comment]
I also need to be able to customize the portal root. I am using components from several libraries that render dialog-type things, and I shuttle them all into a custom `<div id="modal-root">` that appears as the sibling element after my main app content. I need dialogs to be inserted into the modal root in the order that they were opened so that everything stacks in the right order (e.g. nested modals/drawers, popovers on top of modals). I've worked around this by portaling the headless-ui modal content, but then I need to do extra work to remove `inert` where it's inappropriately injected. Super frustrating.

**Comment by Crocsx** [comment]
Came here to post this one to. While there are 99% of case where you want portal on body, there are some cases, specifically due to design compromise, where it is necessary to display a drawer for exemple into a container and this is almost impossible or highly inconvenient to do in css.

Having the ability to define a container for the portal in those rare case would be a big win.

Would be nice to have some maintainer opinion cause it's been 3 years of discussion and this feature is clearly a popular one

**Comment by mikeskydev** [comment]
I would also love a solution, my use case is a WebXR App that uses the [DOM Overlay](https://www.w3.org/TR/webxr-dom-overlays-1/) module, which renders using `top-layer` ([explainer](https://developer.chrome.com/blog/what-is-the-top-layer)). This means that all Dialogs appear underneath the rest of my XR UI...

**Comment by boris-etingof-flarehr** [comment]
Here's the approach that works for me. I render a Headless UI Dialog inside a web component, but when it's teleported outside, the component's shadow-DOM styles no longer apply. To fix this, I observe mutations on document.body. As soon as a headlessui-portal-root element is added, I move it back into the component's shadow root, when it's removed, I clean up the shadow root.

I'm on "@headlessui/react": "1.7.16"

The code lives in the top-level App.tsx, outside the component function.

```
const headlessuiPortalRootId = 'headlessui-portal-root';
const webComponentTagName = 'my-web-component';

const findHeadlessPortalRootNodes = (nodes: NodeList) => {
  return Array.from(nodes).some(
    (n): n is HTMLElement =>
      n instanceof HTMLElement && n.id === headlessuiPortalRootId,
  );
};

(() => {
  const move = (mutations: MutationRecord[]) => {
    const {added, removed} = mutations.reduce<{
      added?: boolean;
      removed?: boolean;
    }>(
      (memo, {addedNodes, removedNodes}) => ({
        added: findHeadlessPortalRootNodes(addedNodes) || memo.added,
        removed: findHeadlessPortalRootNodes(removedNodes) || memo.removed,
      }),
      {},
    );

    const root = document.getElementById(headlessuiPortalRootId);
    const target = document.querySelector(webComponentTagName)?.shadowRoot;

    if (added && root?.firstChild) {
      target?.appendChild(root.firstChild);
    }

    if (removed) {
      const portals = target?.querySelectorAll('[data-headlessui-portal]');
      portals?.forEach(el => el.remove());
    }
  };

  const docObs = new MutationObserver(move);
  docObs.observe(document.body, {childList: true});

  return () => {
    docObs.disconnect();
  };
})();

const App: FC = () => {
   ...
}

export default App;

```
